File: js/src/blob.ts
import { createHash } from "node:crypto";

import { ModalGrpcClient } from "./client";

const blobUploadRetryAttempts = 3;
const blobUploadRetryDelayMs = 300;
const blobDownloadRetryAttempts = 5;
const blobDownloadRetryDelayMs = 100;

async function retryHttpRequest<T>(
  fn: () => Promise<T>,
  attempts: number,
  baseDelayMs: number,
): Promise<T> {
  let delay = baseDelayMs;
  let lastErr: unknown;
  for (let attempt = 0; attempt < attempts; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastErr = err;
      if (attempt < attempts - 1) {
        await new Promise((resolve) => setTimeout(resolve, delay));
        delay *= 2;
      }
    }
  }
  throw lastErr;
}

export async function blobUpload(
  cpClient: ModalGrpcClient,
  data: Uint8Array,
): Promise<string> {
  const contentMd5 = createHash("md5").update(data).digest("base64");
  const contentSha256 = createHash("sha256").update(data).digest("base64");
  const resp = await cpClient.blobCreate({
    contentMd5,
    contentSha256Base64: contentSha256,
    contentLength: data.length,
  });
  if (resp.multipart) {
    throw new Error(
      "Function input size exceeds multipart upload threshold, unsupported by this SDK version",
    );
  } else if (resp.uploadUrl) {
    await retryHttpRequest(
      async () => {
        const uploadResp = await fetch(resp.uploadUrl!, {
          method: "PUT",
          headers: {
            "Content-Type": "application/octet-stream",
            "Content-MD5": contentMd5,
          },
          body: data,
        });
        if (uploadResp.status < 200 || uploadResp.status >= 300) {
          throw new Error(`Failed blob upload: ${uploadResp.statusText}`);
        }
      },
      blobUploadRetryAttempts,
      blobUploadRetryDelayMs,
    );
    // Skip client-side ETag header validation for now (MD5 checksum).
    return resp.blobId;
  } else {
    throw new Error("Missing upload URL in BlobCreate response");
  }
}

export async function blobDownload(
  cpClient: ModalGrpcClient,
  blobId: string,
): Promise<Uint8Array> {
  const resp = await cpClient.blobGet({ blobId });
  return retryHttpRequest(
    async () => {
      const s3resp = await fetch(resp.downloadUrl);
      if (!s3resp.ok) {
        throw new Error(`Failed to download blob: ${s3resp.statusText}`);
      }
      const buf = await s3resp.arrayBuffer();
      return new Uint8Array(buf);
    },
    blobDownloadRetryAttempts,
    blobDownloadRetryDelayMs,
  );
}

// Exported for testing only.
export { retryHttpRequest as _retryHttpRequest };

File: go/blob.go
package modal

import (
	"bytes"
	"context"
	"crypto/md5"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"time"

	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
)

const (
	blobUploadRetryAttempts   = 3
	blobUploadRetryDelay      = 300 * time.Millisecond
	blobDownloadRetryAttempts = 5
	blobDownloadRetryDelay    = 100 * time.Millisecond
)

// retryHTTPRequest retries an HTTP operation with exponential backoff.
func retryHTTPRequest[T any](ctx context.Context, logger *slog.Logger, operation string, attempts int, baseDelay time.Duration, fn func() (T, error)) (T, error) {
	delay := baseDelay
	var zero T
	var lastErr error
	for attempt := range attempts {
		result, err := fn()
		if err == nil {
			return result, nil
		}
		lastErr = err
		if attempt < attempts-1 {
			logger.DebugContext(ctx, fmt.Sprintf("%s failed (attempt %d/%d), retrying",
				operation, attempt+1, attempts), "error", lastErr.Error())
			select {
			case <-ctx.Done():
				return zero, ctx.Err()
			case <-time.After(delay):
			}
			delay *= 2
		}
	}
	return zero, lastErr
}

func blobUpload(ctx context.Context, client pb.ModalClientClient, logger *slog.Logger, data []byte) (string, error) {
	md5sum := md5.Sum(data)
	sha256sum := sha256.Sum256(data)
	contentMd5 := base64.StdEncoding.EncodeToString(md5sum[:])
	contentSha256 := base64.StdEncoding.EncodeToString(sha256sum[:])

	resp, err := client.BlobCreate(ctx, pb.BlobCreateRequest_builder{
		ContentMd5:          contentMd5,
		ContentSha256Base64: contentSha256,
		ContentLength:       int64(len(data)),
	}.Build())
	if err != nil {
		return "", fmt.Errorf("failed to create blob: %w", err)
	}

	switch resp.WhichUploadTypeOneof() {
	case pb.BlobCreateResponse_Multipart_case:
		return "", fmt.Errorf("Function input size exceeds multipart upload threshold, unsupported by this SDK version")

	case pb.BlobCreateResponse_UploadUrl_case:
		return retryHTTPRequest(ctx, logger, "blob upload", blobUploadRetryAttempts, blobUploadRetryDelay, func() (string, error) {
			req, err := http.NewRequestWithContext(ctx, "PUT", resp.GetUploadUrl(), bytes.NewReader(data))
			if err != nil {
				return "", fmt.Errorf("failed to create upload request: %w", err)
			}
			req.Header.Set("Content-Type", "application/octet-stream")
			req.Header.Set("Content-MD5", contentMd5)
			uploadResp, err := http.DefaultClient.Do(req)
			if err != nil {
				return "", fmt.Errorf("failed to upload blob: %w", err)
			}
			defer func() {
				if err := uploadResp.Body.Close(); err != nil {
					logger.DebugContext(ctx, "failed to close upload response body", "error", err.Error())
				}
			}()
			if uploadResp.StatusCode < 200 || uploadResp.StatusCode >= 300 {
				return "", fmt.Errorf("failed blob upload: %s", uploadResp.Status)
			}
			return resp.GetBlobId(), nil
		})

	default:
		return "", fmt.Errorf("missing upload URL in BlobCreate response")
	}
}

// blobDownload downloads a blob by its ID.
func blobDownload(ctx context.Context, client pb.ModalClientClient, logger *slog.Logger, blobID string) ([]byte, error) {
	resp, err := client.BlobGet(ctx, pb.BlobGetRequest_builder{
		BlobId: blobID,
	}.Build())
	if err != nil {
		return nil, err
	}
	return retryHTTPRequest(ctx, logger, "blob download", blobDownloadRetryAttempts, blobDownloadRetryDelay, func() ([]byte, error) {
		req, err := http.NewRequestWithContext(ctx, "GET", resp.GetDownloadUrl(), nil)
		if err != nil {
			return nil, fmt.Errorf("failed to create download request: %w", err)
		}
		s3resp, err := http.DefaultClient.Do(req)
		if err != nil {
			return nil, fmt.Errorf("failed to download blob: %w", err)
		}
		defer func() {
			if err := s3resp.Body.Close(); err != nil {
				logger.DebugContext(ctx, "failed to close download response body", "error", err.Error())
			}
		}()
		if s3resp.StatusCode < 200 || s3resp.StatusCode >= 300 {
			return nil, fmt.Errorf("failed blob download: %s", s3resp.Status)
		}
		buf, err := io.ReadAll(s3resp.Body)
		if err != nil {
			return nil, fmt.Errorf("failed to read blob data: %w", err)
		}
		return buf, nil
	})
}
